### ISSUE TITLE
elisp-koans/contemplate-koans

### ISSUE THREAD

**Issue by matthew-piziak** [jtmoulia/elisp-koans#2]
It would be nice to have a `contemplate-koans` interactive function which loads and runs the ERT tests in the correct order and jumps you to the failing test case. See `comtemplate_koans.py` at https://github.com/gregmalcolm/python_koans.

**Comment by jtmoulia** [comment]
Sorry for missing this @matthew-piziak -- yup, that'd be a huge improvement to the workflow.

Are you interested in implementing this? Otherwise I'm happy to put it on my TODO list.

**Comment by matthew-piziak** [comment]
Thanks @jtmoulia. 

I started hacking together something, take a look:
```
(defun elisp-koans/contemplate ()
  "Run koans in order and jump to next failure or blank."
  (interactive)
  (elisp-koans/load-koan-groups)
  (ert-run-tests-interactively
   `(member
     ,@(-flatten
        (mapcar
         (lambda (group)
           (let* ((file-name (concat (symbol-name group) ".el"))
                  (file-path
                   (concat (file-name-as-directory elisp-koans--groups-directory) file-name)))
             (cdadr (assoc 'define-symbol-props (cdr (assoc file-path load-history))))))
         elisp-koans-groups))))
  (sit-for 2)
  (forward-button 3)
  (let ((failure (button-label (button-at (point)))))
    (push-button)
    (search-forward "___")
    (message "%s has damaged your karma" failure)))
```

Obviously it's messy, but I wanted to see if I could get something working.

Unfortunately I'm finding a lot of undefined symbols. For example `elisp-koans/true-or-false?` isn't defined. How does that work in the following form?

```
(elisp-koans/deftest
 elisp-koans/nil-false-empty-test-or ()
 "`or' can also take multiple arguments"
 (elisp-koans/true-or-false? ____  (or nil nil nil t nil))
 "`or' returns the first non nil value, or nil if there are none."
 (elisp-koans/assert-equal ____ (or nil nil nil))
 (elisp-koans/assert-equal ____ (or 1 2 3 4 5)))
```

**Comment by jtmoulia** [comment]
Awesome @matthew-piziak! I'm running the function, and the workflow is great.

I did make some changes in `master` underneath you, specifically `elisp-koans/load-koan-groups` is now `elisp-koans/load-groups` and there is an `elisp-koans/run-tests` function to run all tests. Including those changes gives you:

```lisp
(defun elisp-koans/contemplate ()
  "Run koans in order and jump to next failure or blank."
  (interactive)
  (elisp-koans/load-groups)
  (elisp-koans/run-tests)
  (sit-for 2)
  (forward-button 3)
  (let ((failure (button-label (button-at (point)))))
    (push-button)
    (search-forward "___")
    (message "%s has damaged your karma" failure)))
```

Otherwise, some additional feedback with comments inline:

```lisp
(defun elisp-koans/contemplate ()
  "Run koans in order and jump to next failure or blank."
  (interactive)
  (elisp-koans/load-groups)
  (elisp-koans/run-tests)
  ;; Suggestion: check for when the tests are complete -- this will retry every 0.2s for 3s.
  ;; Note: This doesn't check for when the koan tests are "Aborted." rather than "Finished."
  (let ((retry-for 3.0)
        (retry-interval 0.2))
    (while (not (re-search-forward "^Finished\.$" nil t))
      (if ((>= 0 retries))
          (error "elisp-koans did not succcessfully complete!"))
      (sit-for retry-interval)
      (setf retries (- retry-for retry-interval))))
  (forward-button 3)
  (let ((failure (button-label (button-at (point)))))
    (push-button)
    ;; Suggestion: match `__`, `___`, or `____`
    (re-search-forward "___?_?")
    ;; Suggestion: This will move the cursor to the beginning of the blank 
    (backward-to-word)
    (forward-char)
    (message "%s has damaged your karma" failure)))

```

Re `elisp-koans/assert-equal`: I went through and removed those broken assertions, the latest commit in `master` has the fix. I should've done this earlier, but I'm putting together a `cheatsheet` branch (currently `wip-cheatsheet`) which contains all of the solutions, and ensures it all works.

**Comment by jtmoulia** [comment]
Awesome @matthew-piziak! I'm running the function, and the workflow is great.

I did make some changes in `master` underneath you, specifically `elisp-koans/load-koan-groups` is now `elisp-koans/load-groups` and there is an `elisp-koans/run-tests` function to run all tests. Including those changes gives you:

```lisp
(defun elisp-koans/contemplate ()
  "Run koans in order and jump to next failure or blank."
  (interactive)
  (elisp-koans/load-groups)
  (elisp-koans/run-tests)
  (sit-for 2)
  (forward-button 3)
  (let ((failure (button-label (button-at (point)))))
    (push-button)
    (search-forward "___")
    (message "%s has damaged your karma" failure)))
```

Otherwise, some additional feedback with comments inline:

```lisp
(defun elisp-koans/contemplate ()
  "Run koans in order and jump to next failure or blank."
  (interactive)
  (elisp-koans/load-groups)
  (elisp-koans/run-tests)
  ;; Suggestion: check for when the tests are complete -- this will retry every 0.2s for 3s.
  ;; Note: This doesn't check for when the koan tests are "Aborted." rather than "Finished."
  (let ((retry-for 3.0)
        (retry-interval 0.2))
    (while (not (re-search-forward "^Finished\.$" nil t))
      (if ((>= 0 retries))
          (error "elisp-koans did not succcessfully complete!"))
      (sit-for retry-interval)
      (setf retries (- retry-for retry-interval))))
  (forward-button 3)
  (let ((failure (button-label (button-at (point)))))
    (push-button)
    ;; Suggestion: match `__`, `___`, or `____`
    (re-search-forward "___?_?")
    ;; Suggestion: This will move the cursor to the beginning of the blank 
    (backward-to-word)
    (forward-char)
    (message "%s has damaged your karma" failure)))
```

I've pushed these changes to the `wip-contemplate` to more easily share.

Re `elisp-koans/assert-equal`: I went through and removed those broken assertions, the latest commit in `master` has the fix. I should've done this earlier, but I'm putting together a `cheatsheet` branch (currently `wip-cheatsheet`) which contains all of the solutions, and ensures it all works.

**Comment by matthew-piziak** [comment]
@jtmoulia That works great!

A note though, the reason I customized the ERT selector in my version is so that the tests run in the correct order. That is, by group, and then in file order.

**Comment by jtmoulia** [comment]
> A note though, the reason I customized the ERT selector in my version is so that the tests run in the correct order. That is, by group, and then in file order.

Ah good to know -- then I'll look into fixing `elisp-koans/run-tests` with the logic in your snippet.

**Comment by jtmoulia** [comment]
Alright @matthew-piziak, I swapped out the runner in `elisp-koans/run-tests` with the logic you provided. Indeed, it's now running the tests in the right order.

**Comment by matthew-piziak** [comment]
@jtmoulia Great! Is anything else left to be done?

**Comment by jtmoulia** [comment]
There's one other issue I ran into while testing: if an assertion fails due to a normal broken assertion (e.g. `(should nil)`) there is no blank `___` to search forward to.

We can work that out over a pull request, though -- are you interested in opening a PR @matthew-piziak ?

**Comment by matthew-piziak** [comment]
I've made a PR here: https://github.com/jtmoulia/elisp-koans/pull/4

I don't know of ERT functionality that will jump to the location of the failing assertion directly. It probably _should_ exist upstream.

**Comment by jtmoulia** [comment]
Closed by #4
